### PR TITLE
Generate report for task once it is done

### PR DESCRIPTION
**Pull request by smthwas**
[TMGP4-211] Create Backend Code
https://sweng894group4.atlassian.net/browse/TMGP4-211

[TMGP4-208] Create Backend Code
https://sweng894group4.atlassian.net/browse/TMGP4-208

For two use cases:
[TMGP4-31] Report on Tasks Completed Behind Schedule
[TMGP4-26] Report on Tasks Completed Ahead of Schedule

When one task is done, the system would generate one report for the task, and send one notification to the user.

The Report would contain if the task is done

ON_TIME: the actual end time is in 10 minutes of the scheduled end time
EARLIER: the actual end time is more than 10 minutes earlier than the scheduled end time
LATER: the actual end time is more than 10 minutes later than the scheduled end time
And the difference between the actual end time and the scheduled end time

The notification would have type as 'task_report', and key as the task's id

The frontend could call two APIs to retrieve the report(s):

GET /report: return all reports for the user
GET /report/task/{TASK_ID}: return the report for the specific task
All the new APIs' document is in https://sweng894group4.atlassian.net/browse/TMGP4-31

### CODE REVIEW

**Comment by smthwas** [comment]
@bashiir @husna Please review this PR. Thanks.

File: BACKEND/userAuth/src/main/java/com/apptime/auth/controller/TaskReportController.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        return new ResponseEntity<>(service.getReports(username), HttpStatus.OK);
+    }
+
+    @GetMapping(value = "/task/{taskId}")
+    public ResponseEntity<TaskReport> getByTask(@PathVariable Long taskId, Authentication authentication) {
+        String username = getUsername(authentication);
+        if (username == null || username.isEmpty()) {
+            return new ResponseEntity<>(HttpStatus.UNAUTHORIZED);
+        }
+        TaskReport taskReport = reportRepository.findByTaskId(taskId);
```
Comment (bashiir) [inline BACKEND/userAuth/src/main/java/com/apptime/auth/controller/TaskReportController.java]: Repository layer shouldn't be used in the controller. if that is the case then what is the point of the service layer.

File: BACKEND/userAuth/src/main/java/com/apptime/auth/model/TaskReport.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+ * This is the PO for TaskReport
+ * Use Case: TMGP4-26, TMGP4-31
+ */
+@Entity
+public class TaskReport {
+    @Id
+    @GeneratedValue
+    private int id;
+
+    @OneToOne
```
Comment (bashiir) [inline BACKEND/userAuth/src/main/java/com/apptime/auth/model/TaskReport.java]: Why we are creating this relationship to task. the report is just  information that we extract from tasks. 
this creates a update /delete cascade problem.

File: BACKEND/userAuth/src/main/java/com/apptime/auth/service/TaskManagerService.java [hunk 3]
```diff
@@ -3,13 +3,13 @@
 import com.apptime.auth.config.TaskStateMachine;
 import com.apptime.auth.model.Task;
 import com.apptime.auth.model.TaskState;
+import com.apptime.auth.repository.TaskReportRepository;
```
Comment (bashiir) [inline BACKEND/userAuth/src/main/java/com/apptime/auth/service/TaskManagerService.java]: I would keep the two service separate to make testing easy. They are parallel. they can both depend on lower level layer. Data access layer.

**Comment by smthwas** [comment]
@bashiir i will fix the issues in the following PR.

**Pull request merged**